controllers/etcdpeer_controller.go

```diff
@@ -16,14 +20,104 @@ type EtcdPeerReconciler struct {
 	Log logr.Logger
 }
 
+func ignoreNotFound(err error) error {
+	if apierrs.IsNotFound(err) {
+		return nil
+	}
+	return err
+}
+
+const (
+	EtcdImage = "quay.io/coreos/etcd:v3.2.27"
+)
+
 // +kubebuilder:rbac:groups=etcd.improbable.io,resources=etcdpeers,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=etcd.improbable.io,resources=etcdpeers/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=apps,resources=replicaset,verbs=get;update;patch;create
+
+func defineReplicaSet(etcdPeer etcdv1alpha1.EtcdPeer) (appsv1.ReplicaSet, error) {
+	var replicas int32 = 1
+
+	return appsv1.ReplicaSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:      make(map[string]string),
+			Annotations: make(map[string]string),
+			Name:        etcdPeer.Name,
+			Namespace:   etcdPeer.Namespace,
+		},
+		Spec: appsv1.ReplicaSetSpec{
+			// This will *always* be 1. Other peers are handled by other EtcdPeers.
+			Replicas: &replicas,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"app":  "etcd",
+					"peer": etcdPeer.Name,
+				},
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"app":  "etcd",
+						"peer": etcdPeer.Name,
+					},
+					Annotations: make(map[string]string),
+					Name:        etcdPeer.Name,
+					Namespace:   etcdPeer.Namespace,
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:  "etcd",
+							Image: EtcdImage,
+						},
+					},
+				},
+			},
+		},
+	}, nil
+}
 
 func (r *EtcdPeerReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
-	_ = r.Log.WithValues("etcdpeer", req.NamespacedName)
+	ctx := context.Background()
+	log := r.Log.WithValues("etcdpeer", req.NamespacedName)
+
+	var etcdPeer etcdv1alpha1.EtcdPeer
+	if err := r.Get(ctx, req.NamespacedName, &etcdPeer); err != nil {
+		log.Error(err, "unable to fetch EtcdPeer")
+		return ctrl.Result{}, ignoreNotFound(err)
+	}
+
+	log.V(1).Info("Found EtcdPeer",
+		"name", etcdPeer.Name,
+		"bootstrapPeers", etcdPeer.Spec.BootstrapPeers)
+
+	var existingReplicaSet appsv1.ReplicaSet
+	err := r.Get(ctx,
+		client.ObjectKey{
+			Namespace: etcdPeer.Namespace,
+			Name:      etcdPeer.Name,
+		},
+		&existingReplicaSet)
+
+	if apierrs.IsNotFound(err) {
+		log.V(1).Info("Replica set does not exist, creating")
+		replicaSet, err := defineReplicaSet(etcdPeer)
+		if err != nil {
+			log.Error(err, "unable to generate ReplicaSet from EtcdPeer")
+			return ctrl.Result{}, err
+		}
+
+		if err := r.Create(ctx, &replicaSet); err != nil {
+			log.Error(err, "unable to create ReplicaSet for EtcdPeer", "replicaSet", replicaSet)
+			return ctrl.Result{}, err
+		}
 
```

Comment:
The creation of the ReplicaSet here should automatically cause this EtcdPeer resource to be re-reconciled.

It's a lot easier to reason about and write controllers that evaluate the current state, make one action, and return in order to make progress. It ensures that you write 'reactive' controllers that are correctly monitoring the current state of the world in order to take actions.

For that reason, I think you should `return ctrl.Result{}, nil` here at the end of the `if apierrs.IsNotFound(err) {` block.

Once the informer observes the replicaset being created, the EtcdPeer will be resynced and this Reconcile method will instead take the next required action.

As an aside, this will also allow you to remove all of the `else` keywords from the code after this block too, which looks clearer and is better go hygiene 😄

Reply:
> It's a lot easier to reason about and write controllers that evaluate the current state, make one action, and return in order to make progress. It ensures that you write 'reactive' controllers that are correctly monitoring the current state of the world in order to take actions.

All onboard with this. :+1:

> For that reason, I think you should return `ctrl.Result{}, nil` here at the end of the `if apierrs.IsNotFound(err) {` block.
> 
> Once the informer observes the replicaset being created, the EtcdPeer will be resynced and this Reconcile method will instead take the next required action.
>
> As an aside, this will also allow you to remove all of the else keywords from the code after this block too, which looks clearer and is better go hygiene 😄

I don't understand. This whole block just does one operation already: creates the replica set:

```go
if apierrs.IsNotFound(err) {
	log.V(1).Info("Replica set does not exist, creating")
	replicaSet, err := defineReplicaSet(etcdPeer)
	if err != nil {
		log.Error(err, "unable to generate ReplicaSet from EtcdPeer")
		return ctrl.Result{}, err
	}

	if err := r.Create(ctx, &replicaSet); err != nil {
		log.Error(err, "unable to create ReplicaSet for EtcdPeer", "replicaSet", replicaSet)
		return ctrl.Result{}, err
	}

} else if err != nil {
	log.Error(err, "unable to query for replica sets")
	return ctrl.Result{}, err
} else {
	log.V(1).Info("Replica set already exists")
}
```

The else chain doesn't *do* anything, it just provides logging if something went wrong. So I'm unclear what you're suggesting I change.

Reply:
If you were to add some additional operation *after* this entire block of code, it would still run this code after creating the ReplicaSet. You're correct that right now it makes no difference, but thought I would raise it now 😄

For an example of structuring your controller in this way, take a look at our acmeorders controller 😄 we recently rewrote it like this, and there's still a few rough edges but it's pretty neat: https://github.com/jetstack/cert-manager/blob/master/pkg/controller/acmeorders/sync.go#L41

Reply:
Ah I see what you're getting at. I've flattened it out and added the return. How does that look @munnerz?